File: mockgcp/mockbigqueryconnection/connection.go
// Copyright 2024 Google LLC
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

package mockbigqueryconnection

import (
	"context"
	"fmt"
	"math/rand"
	"strconv"
	"strings"
	"time"

	"github.com/golang/protobuf/ptypes/empty"
	"google.golang.org/grpc/codes"
	"google.golang.org/grpc/status"
	"google.golang.org/protobuf/proto"
	"google.golang.org/protobuf/types/known/emptypb"

	"github.com/GoogleCloudPlatform/k8s-config-connector/mockgcp/common/projects"
	pb "github.com/GoogleCloudPlatform/k8s-config-connector/mockgcp/generated/mockgcp/cloud/bigquery/connection/v1"
)

type ConnectionV1 struct {
	*MockService
	pb.UnimplementedConnectionServiceServer
}

func (s *ConnectionV1) GetConnection(ctx context.Context, req *pb.GetConnectionRequest) (*pb.Connection, error) {
	name, err := s.parseConnectionName(req.Name)
	if err != nil {
		return nil, err
	}

	fqn := name.String()

	obj := &pb.Connection{}
	if err := s.storage.Get(ctx, fqn, obj); err != nil {
		if status.Code(err) == codes.NotFound {
			return nil, status.Errorf(codes.NotFound, "Not found: Connection %s", fqn)
		}
		return nil, err
	}

	return obj, nil
}

func (s *ConnectionV1) CreateConnection(ctx context.Context, req *pb.CreateConnectionRequest) (*pb.Connection, error) {
	var reqName string
	if req.ConnectionId != "" {
		reqName = req.Parent + "/connections/" + req.ConnectionId
	} else if req.Connection.Name != "" {
		reqName = req.Connection.Name
	} else {
		// reqName = req.Parent + "/connections/" + uuid.New().String()
		// Using fixed UUID to test "acquire" in spec.resourceID. This also fix the dynamic uuid value in the `x-goog-request-params` header.
		reqName = req.Parent + "/connections/" + "71389360-831c-431d-8975-837aee2153be"
	}
	name, err := s.parseConnectionName(reqName)
	if err != nil {
		return nil, err
	}

	fqn := name.String()
	now := time.Now()

	obj := proto.Clone(req.Connection).(*pb.Connection)

	obj.Name = name.String()
	obj.CreationTime = now.Unix()
	obj.LastModifiedTime = now.Unix()

	// Bigqueryconnections supports two types of SA.
	// Primary SA is attached with pre-defined IAM roles, while the delegation SA doesn't include any roles.
	// see https://cloud.google.com/iam/docs/service-agents#bigquery-connection-delegation-service-agent.

	buildDelegationServiceAccountId := func() string {
		letterRunes := []rune("abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ")
		b := make([]rune, 4)
		for i := range b {
			b[i] = letterRunes[rand.Intn(len(letterRunes))]
		}
		return fmt.Sprintf("bqcx-%s-%s@gcp-sa-bigquery-condel.iam.gserviceaccount.com", req.GetParent(), string(b))
	}

	buildPrimaryServiceAccountId := func() string {
		return fmt.Sprintf("service-%s@gcp-sa-bigqueryconnection.iam.gserviceaccount.com", req.GetParent())
	}

	buildGoogleIdentity := func() string {
		letterRunes := []rune("0123456789")
		b := make([]rune, 21)
		for i := range b {
			b[i] = letterRunes[rand.Intn(len(letterRunes))]
		}
		return string(b)
	}

	if _, ok := (req.Connection.Properties).(*pb.Connection_Aws); ok {
		if aws := req.Connection.GetAws(); aws != nil {
			obj.Properties = &pb.Connection_Aws{
				Aws: &pb.AwsProperties{
					AuthenticationMethod: &pb.AwsProperties_AccessRole{
						AccessRole: &pb.AwsAccessRole{
							IamRoleId: aws.GetAccessRole().GetIamRoleId(),
							Identity:  buildGoogleIdentity(),
						},
					},
				},
			}
		}
	}

	if _, ok := (req.Connection.Properties).(*pb.Connection_Azure); ok {
		if azure := req.Connection.GetAzure(); azure != nil {
			obj.Properties = &pb.Connection_Azure{
				Azure: &pb.AzureProperties{
					CustomerTenantId:             azure.GetCustomerTenantId(),
					FederatedApplicationClientId: azure.GetFederatedApplicationClientId(),
					Identity:                     buildGoogleIdentity(),
				},
			}
		}
	}

	if _, ok := (req.Connection.Properties).(*pb.Connection_CloudResource); ok {
		obj.Properties = &pb.Connection_CloudResource{
			CloudResource: &pb.CloudResourceProperties{
				ServiceAccountId: buildDelegationServiceAccountId(),
			},
		}
	}

	if _, ok := (req.Connection.Properties).(*pb.Connection_CloudSql); ok {
		obj.HasCredential = true
		if sql := req.Connection.GetCloudSql(); sql != nil {
			obj.Properties = &pb.Connection_CloudSql{
				CloudSql: &pb.CloudSqlProperties{
					InstanceId:       sql.InstanceId,
					Database:         sql.Database,
					Type:             sql.Type,
					ServiceAccountId: buildPrimaryServiceAccountId(),
				},
			}
		}
	}

	if _, ok := (req.Connection.Properties).(*pb.Connection_CloudSpanner); ok {
		if spanner := req.Connection.GetCloudSpanner(); spanner != nil {
			obj.Properties = &pb.Connection_CloudSpanner{
				CloudSpanner: &pb.CloudSpannerProperties{
					Database:       spanner.Database,
					UseParallelism: spanner.UseParallelism,
					UseDataBoost:   spanner.UseDataBoost,
					MaxParallelism: spanner.MaxParallelism,
					DatabaseRole:   spanner.DatabaseRole,
				},
			}
		}
	}

	if _, ok := (req.Connection.Properties).(*pb.Connection_Spark); ok {
		if spark := req.Connection.GetSpark(); spark != nil {
			obj.Properties = &pb.Connection_Spark{
				Spark: &pb.SparkProperties{
					MetastoreServiceConfig:   spark.GetMetastoreServiceConfig(),
					SparkHistoryServerConfig: spark.GetSparkHistoryServerConfig(),
					ServiceAccountId:         buildDelegationServiceAccountId(),
				},
			}
		}
	}

	if err := s.storage.Create(ctx, fqn, obj); err != nil {
		return nil, err
	}
	return obj, nil
}

func (s *ConnectionV1) UpdateConnection(ctx context.Context, req *pb.UpdateConnectionRequest) (*pb.Connection, error) {
	name, err := s.parseConnectionName(req.Name)
	if err != nil {
		return nil, err
	}

	fqn := name.String()
	now := time.Now()

	obj := &pb.Connection{}
	if err := s.storage.Get(ctx, fqn, obj); err != nil {
		return nil, err
	}

	paths := req.GetUpdateMask().GetPaths()
	for _, path := range paths {
		switch path {
		case "friendlyName":
			obj.FriendlyName = req.GetConnection().GetFriendlyName()
		case "description":
			obj.Description = req.GetConnection().GetDescription()
		default:
			return nil, status.Errorf(codes.InvalidArgument, "update_mask path %q not valid", path)
		}
	}
	obj.LastModifiedTime = now.Unix()

	if _, ok := (req.Connection.Properties).(*pb.Connection_Aws); ok {
		if mod := req.Connection.GetAws(); mod != nil {
			obj.GetAws().GetAccessRole().IamRoleId = mod.GetAccessRole().IamRoleId
		}
	}

	if _, ok := (req.Connection.Properties).(*pb.Connection_CloudSpanner); ok {
		if mod := req.Connection.GetCloudSpanner(); mod != nil {
			obj.GetCloudSpanner().Database = mod.Database
			obj.GetCloudSpanner().UseDataBoost = mod.UseDataBoost
			obj.GetCloudSpanner().UseParallelism = mod.UseParallelism
			obj.GetCloudSpanner().MaxParallelism = mod.MaxParallelism
			obj.GetCloudSpanner().DatabaseRole = mod.DatabaseRole
		}
	}

	if err := s.storage.Update(ctx, fqn, obj); err != nil {
		return nil, err
	}
	return obj, nil
}

func (s *ConnectionV1) DeleteConnection(ctx context.Context, req *pb.DeleteConnectionRequest) (*empty.Empty, error) {
	name, err := s.parseConnectionName(req.Name)
	if err != nil {
		return nil, err
	}

	fqn := name.String()

	oldObj := &pb.Connection{}
	if err := s.storage.Delete(ctx, fqn, oldObj); err != nil {
		return nil, err
	}
	return &emptypb.Empty{}, nil
}

type connectionName struct {
	Project    *projects.ProjectData
	Location   string
	ResourceID string
}

func (n *connectionName) String() string {
	return "projects/" + strconv.FormatInt(n.Project.Number, 10) + "/locations/" + n.Location + "/connections/" + n.ResourceID
}

// parseConnectionName parses a string into a connectionName.
// The expected form is projects/<projectNum>/locations/<location>/connections/<connectionID>
func (s *MockService) parseConnectionName(name string) (*connectionName, error) {
	tokens := strings.Split(name, "/")

	if len(tokens) == 6 && tokens[0] == "projects" && tokens[2] == "locations" && tokens[4] == "connections" {
		project, err := s.Projects.GetProjectByID(tokens[1])
		if err != nil {
			return nil, err
		}

		name := &connectionName{
			Project:    project,
			Location:   tokens[3],
			ResourceID: tokens[5],
		}

		return name, nil
	} else {
		return nil, status.Errorf(codes.InvalidArgument, "name %q is not valid", name)
	}
}

File: python/services/bigqueryconnection/connection.py
# Copyright 2021 Google LLC. All Rights Reserved.
# 
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
# 
#     http://www.apache.org/licenses/LICENSE-2.0
# 
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
from connector import channel
from google3.cloud.graphite.mmv2.services.google.bigquery_connection import (
    connection_pb2,
)
from google3.cloud.graphite.mmv2.services.google.bigquery_connection import (
    connection_pb2_grpc,
)

from typing import List


class Connection(object):
    def __init__(
        self,
        name: str = None,
        friendly_name: str = None,
        description: str = None,
        cloud_sql: dict = None,
        creation_time: int = None,
        last_modified_time: int = None,
        has_credential: bool = None,
        project: str = None,
        location: str = None,
        service_account_file: str = "",
    ):

        channel.initialize()
        self.name = name
        self.friendly_name = friendly_name
        self.description = description
        self.cloud_sql = cloud_sql
        self.project = project
        self.location = location
        self.service_account_file = service_account_file

    def apply(self):
        stub = connection_pb2_grpc.BigqueryconnectionConnectionServiceStub(
            channel.Channel()
        )
        request = connection_pb2.ApplyBigqueryconnectionConnectionRequest()
        if Primitive.to_proto(self.name):
            request.resource.name = Primitive.to_proto(self.name)

        if Primitive.to_proto(self.friendly_name):
            request.resource.friendly_name = Primitive.to_proto(self.friendly_name)

        if Primitive.to_proto(self.description):
            request.resource.description = Primitive.to_proto(self.description)

        if ConnectionCloudSql.to_proto(self.cloud_sql):
            request.resource.cloud_sql.CopyFrom(
                ConnectionCloudSql.to_proto(self.cloud_sql)
            )
        else:
            request.resource.ClearField("cloud_sql")
        if Primitive.to_proto(self.project):
            request.resource.project = Primitive.to_proto(self.project)

        if Primitive.to_proto(self.location):
            request.resource.location = Primitive.to_proto(self.location)

        request.service_account_file = self.service_account_file

        response = stub.ApplyBigqueryconnectionConnection(request)
        self.name = Primitive.from_proto(response.name)
        self.friendly_name = Primitive.from_proto(response.friendly_name)
        self.description = Primitive.from_proto(response.description)
        self.cloud_sql = ConnectionCloudSql.from_proto(response.cloud_sql)
        self.creation_time = Primitive.from_proto(response.creation_time)
        self.last_modified_time = Primitive.from_proto(response.last_modified_time)
        self.has_credential = Primitive.from_proto(response.has_credential)
        self.project = Primitive.from_proto(response.project)
        self.location = Primitive.from_proto(response.location)

    def delete(self):
        stub = connection_pb2_grpc.BigqueryconnectionConnectionServiceStub(
            channel.Channel()
        )
        request = connection_pb2.DeleteBigqueryconnectionConnectionRequest()
        request.service_account_file = self.service_account_file
        if Primitive.to_proto(self.name):
            request.resource.name = Primitive.to_proto(self.name)

        if Primitive.to_proto(self.friendly_name):
            request.resource.friendly_name = Primitive.to_proto(self.friendly_name)

        if Primitive.to_proto(self.description):
            request.resource.description = Primitive.to_proto(self.description)

        if ConnectionCloudSql.to_proto(self.cloud_sql):
            request.resource.cloud_sql.CopyFrom(
                ConnectionCloudSql.to_proto(self.cloud_sql)
            )
        else:
            request.resource.ClearField("cloud_sql")
        if Primitive.to_proto(self.project):
            request.resource.project = Primitive.to_proto(self.project)

        if Primitive.to_proto(self.location):
            request.resource.location = Primitive.to_proto(self.location)

        response = stub.DeleteBigqueryconnectionConnection(request)

    @classmethod
    def list(self, project, location, service_account_file=""):
        stub = connection_pb2_grpc.BigqueryconnectionConnectionServiceStub(
            channel.Channel()
        )
        request = connection_pb2.ListBigqueryconnectionConnectionRequest()
        request.service_account_file = service_account_file
        request.Project = project

        request.Location = location

        return stub.ListBigqueryconnectionConnection(request).items

    def to_proto(self):
        resource = connection_pb2.BigqueryconnectionConnection()
        if Primitive.to_proto(self.name):
            resource.name = Primitive.to_proto(self.name)
        if Primitive.to_proto(self.friendly_name):
            resource.friendly_name = Primitive.to_proto(self.friendly_name)
        if Primitive.to_proto(self.description):
            resource.description = Primitive.to_proto(self.description)
        if ConnectionCloudSql.to_proto(self.cloud_sql):
            resource.cloud_sql.CopyFrom(ConnectionCloudSql.to_proto(self.cloud_sql))
        else:
            resource.ClearField("cloud_sql")
        if Primitive.to_proto(self.project):
            resource.project = Primitive.to_proto(self.project)
        if Primitive.to_proto(self.location):
            resource.location = Primitive.to_proto(self.location)
        return resource


class ConnectionCloudSql(object):
    def __init__(
        self,
        instance_id: str = None,
        database: str = None,
        type: str = None,
        credential: dict = None,
    ):
        self.instance_id = instance_id
        self.database = database
        self.type = type
        self.credential = credential

    @classmethod
    def to_proto(self, resource):
        if not resource:
            return None

        res = connection_pb2.BigqueryconnectionConnectionCloudSql()
        if Primitive.to_proto(resource.instance_id):
            res.instance_id = Primitive.to_proto(resource.instance_id)
        if Primitive.to_proto(resource.database):
            res.database = Primitive.to_proto(resource.database)
        if ConnectionCloudSqlTypeEnum.to_proto(resource.type):
            res.type = ConnectionCloudSqlTypeEnum.to_proto(resource.type)
        if ConnectionCloudSqlCredential.to_proto(resource.credential):
            res.credential.CopyFrom(
                ConnectionCloudSqlCredential.to_proto(resource.credential)
            )
        else:
            res.ClearField("credential")
        return res

    @classmethod
    def from_proto(self, resource):
        if not resource:
            return None

        return ConnectionCloudSql(
            instance_id=Primitive.from_proto(resource.instance_id),
            database=Primitive.from_proto(resource.database),
            type=ConnectionCloudSqlTypeEnum.from_proto(resource.type),
            credential=ConnectionCloudSqlCredential.from_proto(resource.credential),
        )


class ConnectionCloudSqlArray(object):
    @classmethod
    def to_proto(self, resources):
        if not resources:
            return resources
        return [ConnectionCloudSql.to_proto(i) for i in resources]

    @classmethod
    def from_proto(self, resources):
        return [ConnectionCloudSql.from_proto(i) for i in resources]


class ConnectionCloudSqlCredential(object):
    def __init__(self, username: str = None, password: str = None):
        self.username = username
        self.password = password

    @classmethod
    def to_proto(self, resource):
        if not resource:
            return None

        res = connection_pb2.BigqueryconnectionConnectionCloudSqlCredential()
        if Primitive.to_proto(resource.username):
            res.username = Primitive.to_proto(resource.username)
        if Primitive.to_proto(resource.password):
            res.password = Primitive.to_proto(resource.password)
        return res

    @classmethod
    def from_proto(self, resource):
        if not resource:
            return None

        return ConnectionCloudSqlCredential(
            username=Primitive.from_proto(resource.username),
            password=Primitive.from_proto(resource.password),
        )


class ConnectionCloudSqlCredentialArray(object):
    @classmethod
    def to_proto(self, resources):
        if not resources:
            return resources
        return [ConnectionCloudSqlCredential.to_proto(i) for i in resources]

    @classmethod
    def from_proto(self, resources):
        return [ConnectionCloudSqlCredential.from_proto(i) for i in resources]


class ConnectionCloudSqlTypeEnum(object):
    @classmethod
    def to_proto(self, resource):
        if not resource:
            return resource
        return connection_pb2.BigqueryconnectionConnectionCloudSqlTypeEnum.Value(
            "BigqueryconnectionConnectionCloudSqlTypeEnum%s" % resource
        )

    @classmethod
    def from_proto(self, resource):
        if not resource:
            return resource
        return connection_pb2.BigqueryconnectionConnectionCloudSqlTypeEnum.Name(
            resource
        )[len("BigqueryconnectionConnectionCloudSqlTypeEnum") :]


class Primitive(object):
    @classmethod
    def to_proto(self, s):
        if not s:
            return ""
        return s

    @classmethod
    def from_proto(self, s):
        return s
